File: backend-csharp/TheGameIsOn.API/Services/MapGeneratorService.cs
using TheGameIsOn.API.Models;

namespace TheGameIsOn.API.Services;

public record MapGenerationOptions
{
    public int MinNodes { get; init; } = 20;
    public int MaxNodes { get; init; } = 24;
    public int MinLayers { get; init; } = 5;
    public int MaxLayers { get; init; } = 7;
}

public interface IMapGeneratorService
{
    Graph Generate(MapGenerationOptions? options = null);
}

public class MapGeneratorService : IMapGeneratorService
{
    private static readonly MapArea[] AreaProgression =
        { MapArea.Forest, MapArea.Dungeon, MapArea.Ruins, MapArea.Volcano };

    private static readonly Dictionary<string, string> NodeIcons = new()
    {
        ["start"]       = "🏁",
        ["end"]         = "🏁",
        ["battle"]      = "⚔️",
        ["hard battle"] = "💀",
        ["new object"]  = "🪄",
        ["power up"]    = "⚡",
        ["treasure"]    = "🎁",
        ["rest"]        = "🛌"
    };

    private MapArea AreaForLayer(int layerIndex, int totalIntermediateLayers)
    {
        var bucket = Math.Min(
            (int)Math.Floor((double)layerIndex / totalIntermediateLayers * AreaProgression.Length),
            AreaProgression.Length - 1);
        return AreaProgression[bucket];
    }

    public Graph Generate(MapGenerationOptions? options = null)
    {
        options ??= new MapGenerationOptions();

        var totalNodes = RandomInt(options.MinNodes, options.MaxNodes);
        var layerCount = RandomInt(options.MinLayers, options.MaxLayers);
        var intermediateNodes = totalNodes - 2;
        var layerSizes = DistributeNodes(intermediateNodes, layerCount - 2);
        var intermediateLayers = layerCount - 2;
        var layers = new List<List<NodeDefinition>>();

        // Start
        layers.Add(new List<NodeDefinition>
        {
            BuildNode("start", "Start", "start", new NodeLayout(50, 0))
        });

        var nodeIndex = 1;
        for (var layer = 1; layer < layerCount - 1; layer++)
        {
            var count = layerSizes[layer - 1];
            var y = Math.Round((double)layer * 100 / (layerCount - 1));
            var area = AreaForLayer(layer - 1, intermediateLayers);
            var layerNodes = new List<NodeDefinition>();

            for (var index = 0; index < count; index++)
            {
                var eventType = RandomEventType();
                layerNodes.Add(BuildNode(
                    $"node-{nodeIndex}",
                    $"{NodeTitle(eventType)} {nodeIndex + 1}",
                    eventType,
                    PositionForLayer(index, count, y),
                    area));
                nodeIndex++;
            }

            layers.Add(layerNodes);
        }

        // End
        layers.Add(new List<NodeDefinition>
        {
            BuildNode("end", "End", "end", new NodeLayout(50, 100))
        });

        var edges = new List<GraphEdge>();
        for (var layer = 0; layer < layers.Count - 1; layer++)
        {
            var sourceLayer = layers[layer];
            var targetLayer = layers[layer + 1];

            foreach (var source in sourceLayer)
            {
                var targets = PickNearestTargets(source, targetLayer, 1, Math.Min(2, targetLayer.Count));
                foreach (var target in targets)
                    edges.Add(new GraphEdge { From = source.Id, To = target.Id });
            }

            foreach (var target in targetLayer)
            {
                var hasIncoming = edges.Any(e => e.To == target.Id);
                if (!hasIncoming)
                {
                    var source = RandomPick(sourceLayer);
                    edges.Add(new GraphEdge { From = source.Id, To = target.Id });
                }
            }
        }

        AddExtraConnections(layers, edges);

        var nodes = layers.SelectMany(l => l).ToList();
        return new Graph { Nodes = nodes, Edges = edges };
    }

    // -------------------------------------------------------------------------

    private static NodeDefinition BuildNode(string id, string title, string eventType,
        NodeLayout layout, MapArea? area = null)
    {
        return new NodeDefinition
        {
            Id = id,
            Title = title,
            Icon = NodeIcons.GetValueOrDefault(eventType, "◯"),
            Layout = layout,
            Event = new NodeEvent
            {
                Type = eventType == "start" || eventType == "end" ? eventType : eventType,
                Area = area
            }
        };
    }

    private static string NodeTitle(string eventType) => eventType switch
    {
        "battle"      => "Battle",
        "hard battle" => "Hard Battle",
        "new object"  => "New Object",
        "power up"    => "Power Up",
        "treasure"    => "Treasure",
        "rest"        => "Rest",
        _             => "Node"
    };

    private static string RandomEventType()
    {
        var choice = Random.Shared.NextDouble();
        if (choice < 0.55) return "battle";
        if (choice < 0.75) return "treasure";
        if (choice < 0.85) return "rest";
        if (choice < 0.93) return "hard battle";
        if (choice < 0.97) return "new object";
        return "power up";
    }

    private static NodeLayout PositionForLayer(int index, int count, double y)
    {
        var x = count == 1 ? 50.0 : 15.0 + 70.0 * index / (count - 1);
        return new NodeLayout(x, y);
    }

    private static List<int> DistributeNodes(int total, int layers)
    {
        var baseCount = total / layers;
        var remainder = total % layers;
        return Enumerable.Range(0, layers)
            .Select(i => baseCount + (i < remainder ? 1 : 0))
            .ToList();
    }

    private static int RandomInt(int min, int max) =>
        Random.Shared.Next(min, max + 1);

    private static T RandomPick<T>(List<T> items) =>
        items[Random.Shared.Next(items.Count)];

    private static void AddExtraConnections(List<List<NodeDefinition>> layers, List<GraphEdge> edges)
    {
        for (var layerIndex = 0; layerIndex < layers.Count - 2; layerIndex++)
        {
            var sourceLayer = layers[layerIndex];
            var targetLayer = layers[layerIndex + 2];

            foreach (var source in sourceLayer)
            {
                if (Random.Shared.NextDouble() < 0.18)
                {
                    var candidates = PickNearestTargets(source, targetLayer, 1, Math.Min(2, targetLayer.Count));
                    if (candidates.Count > 0)
                        edges.Add(new GraphEdge { From = source.Id, To = candidates[0].Id });
                }
            }
        }
    }

    private static List<NodeDefinition> PickNearestTargets(
        NodeDefinition source, List<NodeDefinition> targets, int min, int max)
    {
        var sx = source.Layout?.X ?? 0;
        var sorted = targets.OrderBy(t => Math.Abs((t.Layout?.X ?? 0) - sx)).ToList();
        var k = Math.Max(min, Math.Min(max, sorted.Count));
        return sorted.Take(k).ToList();
    }
}

File: backend/src/services/map-generator.service.ts
import { NodeDefinition } from '../models/node';
import { NodeEventType, MapArea } from '../models/node-event';

interface MapGenerationOptions {
  minNodes?: number;
  maxNodes?: number;
  minLayers?: number;
  maxLayers?: number;
}

const nodeIcons: Record<string, string> = {
  start: '🏁',
  end: '🏁',
  battle: '⚔️',
  'hard battle': '💀',
  'new object': '🪄',
  'power up': '⚡',
  treasure: '🎁',
  rest: '🛌'
};

export class MapGeneratorService {
    /**
   * Map areas assigned in order of progression.
   * Layer depth is divided into equal buckets; each bucket maps to one zone.
   */
  private static readonly AREA_PROGRESSION: MapArea[] = [
    'forest',
    'dungeon',
    'ruins',
    'volcano',
  ];

  /** Returns the zone for an intermediate layer index (0-based, excludes start/end). */
  private areaForLayer(layerIndex: number, totalIntermediateLayers: number): MapArea {
    const areas  = MapGeneratorService.AREA_PROGRESSION;
    const bucket = Math.min(
      Math.floor((layerIndex / totalIntermediateLayers) * areas.length),
      areas.length - 1
    );
    return areas[bucket];
  }

  generate(options: MapGenerationOptions = {}): { nodes: NodeDefinition[]; edges: Array<{ from: string; to: string }> } {
    const totalNodes = this.randomInt(options.minNodes ?? 20, options.maxNodes ?? 24);
    const layerCount = this.randomInt(options.minLayers ?? 5, options.maxLayers ?? 7);
    const intermediateNodes = totalNodes - 2;
    const layerSizes = this.distributeNodes(intermediateNodes, layerCount - 2);
    const intermediateLayers = layerCount - 2; // excludes start + end
    const layers: NodeDefinition[][] = [];

    layers.push([this.buildNode('start', 'Start', 'start', { x: 50, y: 0 })]);

    let nodeIndex = 1;
    for (let layer = 1; layer < layerCount - 1; layer++) {
      const count        = layerSizes[layer - 1];
      const y            = Math.round((layer * 100) / (layerCount - 1));
      const area         = this.areaForLayer(layer - 1, intermediateLayers);
      const layerNodes   = Array.from({ length: count }, (_, index) => {
        const eventType = this.randomEventType();
        return this.buildNode(
          `node-${nodeIndex++}`,
          `${this.nodeTitle(eventType)} ${nodeIndex}`,
          eventType,
          this.positionForLayer(index, count, y),
          area
        );
      });
      layers.push(layerNodes);
    }

    layers.push([this.buildNode('end', 'End', 'end', { x: 50, y: 100 })]);

    const edges: Array<{ from: string; to: string }> = [];
    for (let layer = 0; layer < layers.length - 1; layer++) {
      const sourceLayer = layers[layer];
      const targetLayer = layers[layer + 1];

      // prefer connecting to nearby nodes (by x position)
      sourceLayer.forEach(source => {
        const targets = this.pickNearestTargets(source, targetLayer, 1, Math.min(2, targetLayer.length));
        targets.forEach(target => edges.push({ from: source.id, to: target.id }));
      });

      // ensure every target has at least one incoming edge
      targetLayer.forEach(target => {
        const incoming = edges.some(edge => edge.to === target.id);
        if (!incoming) {
          const source = this.randomPick(sourceLayer);
          edges.push({ from: source.id, to: target.id });
        }
      });
    }

    this.addExtraConnections(layers, edges);

    const nodes = layers.flat();
    return { nodes, edges };
  }

    private buildNode(
    id: string,
    title: string,
    eventType: string,
    layout: { x: number; y: number },
    area?: MapArea
  ): NodeDefinition {
    return {
      id,
      title,
      icon: nodeIcons[eventType] || '◯',
      layout,
      event: {
        type: eventType === 'start' ? 'start' : eventType === 'end' ? 'end' : eventType,
        ...(area ? { area } : {}),
      }
    };
  }

  private nodeTitle(eventType: NodeEventType): string {
    switch (eventType) {
      case 'battle':
        return 'Battle';
      case 'hard battle':
        return 'Hard Battle';
      case 'new object':
        return 'New Object';
      case 'power up':
        return 'Power Up';
      case 'treasure':
        return 'Treasure';
      case 'rest':
        return 'Rest';
      default:
        return 'Node';
    }
  }

  private randomEventType(): NodeEventType {
    const choice = Math.random();
    if (choice < 0.55) return 'battle';
    if (choice < 0.75) return 'treasure';
    if (choice < 0.85) return 'rest';
    if (choice < 0.93) return 'hard battle';
    if (choice < 0.97) return 'new object';
    return 'power up';
  }

  private positionForLayer(index: number, count: number, y: number) {
    const x = count === 1 ? 50 : 15 + (70 * index) / (count - 1);
    return { x, y };
  }

  private distributeNodes(total: number, layers: number) {
    const base = Math.floor(total / layers);
    const remainder = total % layers;
    return Array.from({ length: layers }, (_, index) => base + (index < remainder ? 1 : 0));
  }

  private randomInt(min: number, max: number) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  private randomSubset<T>(items: T[], minItems: number, maxItems: number): T[] {
    const count = this.randomInt(minItems, maxItems);
    const copy = [...items];
    const results: T[] = [];
    while (results.length < count && copy.length) {
      const index = this.randomInt(0, copy.length - 1);
      results.push(copy.splice(index, 1)[0]);
    }
    return results;
  }

  private randomPick<T>(items: T[]): T {
    const index = this.randomInt(0, items.length - 1);
    return items[index];
  }

  private addExtraConnections(layers: NodeDefinition[][], edges: Array<{ from: string; to: string }>) {
    for (let layerIndex = 0; layerIndex < layers.length - 2; layerIndex++) {
      const sourceLayer = layers[layerIndex];
      const targetLayer = layers[layerIndex + 2];
      sourceLayer.forEach(source => {
        if (Math.random() < 0.18) {
          // prefer nearby targets for skip-layer connections to avoid long cross-links
          const candidates = this.pickNearestTargets(source, targetLayer, 1, Math.min(2, targetLayer.length));
          if (candidates.length > 0) {
            const target = candidates[0];
            edges.push({ from: source.id, to: target.id });
          }
        }
      });
    }
  }

  private pickNearestTargets(source: NodeDefinition, targets: NodeDefinition[], min: number, max: number) {
    const sorted = targets.slice().sort((a, b) => {
      const ax = a.layout?.x ?? 0;
      const bx = b.layout?.x ?? 0;
      const sx = source.layout?.x ?? 0;
      return Math.abs(ax - sx) - Math.abs(bx - sx);
    });
    const k = Math.max(min, Math.min(max, sorted.length));
    return sorted.slice(0, k);
  }
}
